File: 144_binary-tree-preorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void PreStore(vector<int>& cur, TreeNode* root)
    {
        if (!root)
            return;
        cur.push_back(root->val);
        PreStore(cur, root->left);
        PreStore(cur, root->right);
    }
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> ret{};
        PreStore(ret, root);
        return ret;
    }
};

File: Tree/144_binary-tree-preorder-traversal.java
/*
Pattern: Binary Tree + DFS

Approach:
- Root
- Left
- Right

Key Insight:
- Preorder = Root → Left → Right.

Complexity:
- Time: O(n)
- Space: O(h)

Takeaway:
- Preorder → Root first.
*/
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private List<Integer> Tree = new ArrayList<>();
    public List<Integer> preorderTraversal(TreeNode root) {
        if (root == null)
            return Tree;
        Tree.add(root.val);
        preorderTraversal(root.left);
        preorderTraversal(root.right);
        return Tree;
    }
}
